admin-console/src/main/java/com/agiletec/apsadmin/portal/PageTreeAction.java

```diff
@@ -19,183 +19,266 @@
+            this.addActionError(this.getText("error.page.selectPageParent"));
+            return "pageTree";
+        }
+        if (parent.getCode().equals(currentPage.getParentCode())) {
+            _logger.debug("trying to move a node under it's own parent..");
+            this.addActionError(this.getText("error.page.move.parentUnderChild.notAllowed"));
+            return "pageTree";
+        }
+        if (parent.isChildOf(selectedNode)) {
+            List<String> args = new ArrayList<String>();
+            args.add(parent.getCode());
+            args.add(selectedNode);
+            this.addActionError(this.getText("error.page.move.parentUnderChild.notAllowed"));
+            return "pageTree";
+        }
+        //group check
+        if (!currentPage.getGroup().equals(Group.FREE_GROUP_NAME)
+                && !currentPage.getGroup().equals(parent.getGroup())) {
+            this.addActionError(this.getText("error.page.move.differentRestriction.notAllowed"));
+            return "pageTree";
+        }
+        return null;
+    }
+
+    public String moveWidgetUp() {
+        return this.moveWidget(this.getFrame() - 1);
+    }
+
+    public String moveWidgetDown() {
+        return this.moveWidget(this.getFrame() + 1);
+    }
+
+    protected String moveWidget(int destFrame) {
+        try {
+            int frame = this.getFrame();
+            this.getPageManager().moveWidget(this.getSelectedNode(), frame, destFrame);
+        } catch (Throwable t) {
+            _logger.error("error in moveWidget", t);
+            return FAILURE;
+        }
+        return SUCCESS;
+    }
+
+    public String copy() {
+        String selectedNode = this.getSelectedNode();
+        String check = this.checkSelectedNode(selectedNode);
+        if (null != check) {
+            return check;
+        }
+        //FIXME RICORDARSI DELL'AREA DI NOTIFICA QUANDO PASSEREMO ALL'ULTIMO LAYER... se Ã¨ possibile mettere un'indicazione della pagina copiata
+        this.setCopyingPageCode(selectedNode);
+        return SUCCESS;
+    }
+
+    protected String checkSelectedNode(String selectedNode) {
+        if (null == selectedNode || selectedNode.trim().length() == 0) {
+            this.addActionError(this.getText("error.page.noSelection"));
+            return "pageTree";
+        }
+        if (AbstractPortalAction.VIRTUAL_ROOT_CODE.equals(selectedNode)) {
+            this.addActionError(this.getText("error.page.virtualRootSelected"));
+            return "pageTree";
+        }
+        IPage selectedPage = this.getPageManager().getDraftPage(selectedNode);
+        if (null == selectedPage) {
+            this.addActionError(this.getText("error.page.selectedPage.null"));
+            return "pageTree";
+        }
+        if (!this.isUserAllowed(selectedPage)) {
+            this.addActionError(this.getText("error.page.userNotAllowed"));
+            return "pageTree";
+        }
+        return null;

```

Comment:
why there is no logging on this method? btw the body of each if is the same one used in the method above: shouldn't we definitely move the body to its own private? DRY